src/core/meshcop/border_agent.cpp

```diff
@@ -317,6 +346,227 @@ template <> void BorderAgent::HandleTmf<kUriRelayRx>(Coap::Message &aMessage, co
+#elif OPENTHREAD_CONFIG_THREAD_VERSION == OT_THREAD_VERSION_1_3_1
+    static const char kThreadVersionString[] = "1.3.1";
+#elif OPENTHREAD_CONFIG_THREAD_VERSION == OT_THREAD_VERSION_1_4
+    static const char kThreadVersionString[] = "1.4.0";
+#endif
+
+    Error                  error       = kErrorNone;
+    Message               *message     = Get<MessagePool>().Allocate(Message::kTypeOther);
+    const NetworkName     &networkName = Get<NetworkNameManager>().GetNetworkName();
+    const ExtendedPanId   &extPanId    = Get<ExtendedPanIdManager>().GetExtPanId();
+    const Mac::ExtAddress &extAddr     = Get<Mac::Mac>().GetExtAddress();
+    uint32_t               partitionId = Get<Mle::MleRouter>().GetLeaderData().GetPartitionId();
+#if OPENTHREAD_CONFIG_BORDER_AGENT_ID_ENABLE
+    Id id;
+#endif
+    StateBitmap   state;
+    int32_t       stateUint32;
+    Dns::TxtEntry txtEntry;
+
+    VerifyOrExit(message != nullptr, error = kErrorNoBufs);
+
+#if OPENTHREAD_CONFIG_BORDER_AGENT_ID_ENABLE
+    if (Get<BorderAgent>().GetId(id) == kErrorNone)
+    {
+        txtEntry.Init("id", id.mId, sizeof(id));
+        SuccessOrExit(error = txtEntry.AppendTo(*message));
+    }
+#endif
+    txtEntry.Init("nn", reinterpret_cast<const uint8_t *>(networkName.GetAsCString()),
+                  StringLength(networkName.GetAsCString(), sizeof(networkName)));
+    SuccessOrExit(error = txtEntry.AppendTo(*message));
+    txtEntry.Init("xp", extPanId.m8, sizeof(extPanId));
+    SuccessOrExit(error = txtEntry.AppendTo(*message));
+    txtEntry.Init("tv", reinterpret_cast<const uint8_t *>(kThreadVersionString), sizeof(kThreadVersionString));
+    SuccessOrExit(error = txtEntry.AppendTo(*message));
+    txtEntry.Init("xa", extAddr.m8, sizeof(extAddr));
+    SuccessOrExit(error = txtEntry.AppendTo(*message));
+
+    state = GetStateBitmap();
+#if OPENTHREAD_CONFIG_BORDER_AGENT_EPHEMERAL_KEY_ENABLE
+    state.mEpskcSupported = Get<BorderAgent::EphemeralKeyManager>().GetState() != kStateDisabled;
+#endif
+    stateUint32 = BigEndian::HostSwap32(state.ToUint32());
+    txtEntry.Init("sb", reinterpret_cast<uint8_t *>(&stateUint32), sizeof(stateUint32));
+    SuccessOrExit(error = txtEntry.AppendTo(*message));
+
+    if (state.mThreadIfStatus == kThreadIfStatusActive)
+    {
+        txtEntry.Init("pt", reinterpret_cast<uint8_t *>(&partitionId), sizeof(partitionId));
+        SuccessOrExit(error = txtEntry.AppendTo(*message));
+
+        SuccessOrExit(error = AppendActiveTimestampTxtEntry(*message));
+    }
+
+#if OPENTHREAD_FTD && OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE
+    SuccessOrExit(error = AppendBbrTxtEntry(state, *message));
+#endif
+#if OTBR_ENABLE_BORDER_ROUTING
+    SuccessOrExit(error = AppendOmrTxtEntry(*message));
+#endif
+
+    VerifyOrExit(message->GetLength() <= aLength, error = kErrorNoBufs);
+    SuccessOrExit(error = message->Read(0, aOutput, message->GetLength()));
+    aLength = message->GetLength();
```

Comment:
We allocate a `Message` where the TXT entries are appended, and then read the content into a buffer at the end.

We can simplify and optimize this to directly append the TXT entry into a buffer (no need to allocate message).
- We have `Appender` class which acts as a wrapper over either a `Message` or a data buffer .
- `TxtEntry` already uses `Appender` and allows us to encode entry directly into a buffer. 
-  The `TxtEntry::AppendTo(Appender &aAppender) const` method is currently `private` but we can make it `public` and then use it here.

Reply:
Updated to use the `Appender` approach.